File: macos_arm64/Sources/RimeoAgentMac/Config/AppConfig.swift
import Foundation
import Darwin

final class AppConfig {
    static let shared = AppConfig()

    #if REVIEW
    let appName = "Rimeo Agent (Review)"
    #else
    let appName = "Rimeo Desktop Agent"
    #endif
    let version: String
    let buildNumber: String
    let releaseTag: String
    let displayVersion: String
    #if REVIEW
    let port: UInt16 = 8042
    // Separate cloudflared config dir so the review agent provisions its OWN
    // named tunnel and runs concurrently with the main agent (shared
    // ~/.cloudflared would make the two fight over one tunnel).
    let cloudflaredDir = "\(NSHomeDirectory())/.cloudflared-review"
    #else
    let port: UInt16 = 8000
    let cloudflaredDir = "\(NSHomeDirectory())/.cloudflared"
    #endif
    let rimeoAppURL = "https://rimeo.app"
    let githubRepo  = "ilokhrimenko-lab/rimeo-agent"

    let baseDir:      URL
    let cacheDir:     URL
    let dataFile:     URL
    let logFile:      URL
    let analysisFile: URL

    private(set) var agentID:  String
    private(set) var xmlPath:  String = ""
    private(set) var dbPath:   String
    /// Whether the Rekordbox master.db source is enabled (Library toggle). Default on.
    private(set) var dbSourceEnabled:  Bool = true
    /// Whether the exported XML source is enabled (Library toggle). Default on.
    private(set) var xmlSourceEnabled: Bool = true
    /// Runtime review mode: the agent serves the bundled 15-track royalty-free
    /// demo library instead of the real Rekordbox. Activated when signed in as the
    /// demo/review account (demo@rimeo.app). See activateReviewMode().
    private(set) var reviewMode: Bool = false

    private let queue = DispatchQueue(label: "rimeo.config", qos: .utility)

    private init() {
        let buildInfo = Self.loadBuildInfo()
        version = buildInfo.version
        buildNumber = buildInfo.buildNumber
        releaseTag = buildInfo.releaseTag
        displayVersion = Self.makeDisplayVersion(version: buildInfo.version, buildNumber: buildInfo.buildNumber)

        let appSupport = FileManager.default.urls(
            for: .applicationSupportDirectory, in: .userDomainMask)[0]
        #if REVIEW
        baseDir      = appSupport.appendingPathComponent("RimeoAgentReview")
        #else
        baseDir      = appSupport.appendingPathComponent("RimeoAgent")
        #endif
        cacheDir     = baseDir.appendingPathComponent("cache")
        dataFile     = baseDir.appendingPathComponent("rimo_data.json")
        logFile      = baseDir.appendingPathComponent("agent.log")
        analysisFile = baseDir.appendingPathComponent("analysis_data.json")
        #if REVIEW
        // Review build never touches the real Rekordbox master.db — XML-only
        // (royalty-free demo library via RIMEO_XML_PATH).
        dbPath       = ""
        #else
        dbPath       = Self.detectDBPath()
        #endif

        try? FileManager.default.createDirectory(at: baseDir,  withIntermediateDirectories: true)
        Self.createCacheDir(cacheDir)

        // Persistent agent ID
        let idFile = baseDir.appendingPathComponent("agent_id")
        if let s = try? String(contentsOf: idFile, encoding: .utf8).trimmingCharacters(in: .whitespacesAndNewlines),
           !s.isEmpty {
            agentID = s
        } else {
            agentID = UUID().uuidString
            try? agentID.write(to: idFile, atomically: true, encoding: .utf8)
        }

        // Load paths from .env
        let envFile = baseDir.appendingPathComponent(".env")
        if let content = try? String(contentsOf: envFile, encoding: .utf8) {
            for line in content.components(separatedBy: .newlines) {
                let kv = line.components(separatedBy: "=")
                guard kv.count >= 2 else { continue }
                let key = kv[0].trimmingCharacters(in: .whitespaces)
                let value = kv.dropFirst().joined(separator: "=")
                    .trimmingCharacters(in: .whitespacesAndNewlines)
                if key == "RIMEO_XML_PATH" {
                    xmlPath = value
                } else if key == "RIMEO_DB_PATH", !value.isEmpty {
                    // Review build is XML-only and ignores any DB path.
                    #if !REVIEW
                    dbPath = value
                    #endif
                } else if key == "RIMEO_DB_ENABLED", !value.isEmpty {
                    dbSourceEnabled = (value as NSString).boolValue
                } else if key == "RIMEO_XML_ENABLED", !value.isEmpty {
                    xmlSourceEnabled = (value as NSString).boolValue
                }
            }
        }

        // Restored review session: if the persisted XML points at our generated
        // review library, stay in review mode across restarts.
        if xmlPath.hasSuffix("review_active.xml") { reviewMode = true }
    }

    /// Idempotently (re)create the cache directory. Must be called at startup AND
    /// before every cache write: the dir lives under Application Support and can be
    /// purged at runtime (disk cleaner / manual clear) while the agent keeps running.
    /// When it's gone, ffmpeg output fails with "No such file or directory" and
    /// /stream 503s every track that needs conversion (AIFF→WAV, hi-res→16-bit),
    /// while plain 16-bit tracks still stream — a ragged, silent breakage. See
    /// bug_reports #89 (2026-08-24). Creating the dir once at startup is not enough.
    @discardableResult
    func ensureCacheDir() -> Bool {
        return Self.createCacheDir(cacheDir)
    }

    /// Silent (no logging): this runs inside `init`, and the file logger's own
    /// init reads `AppConfig.shared` — touching `logger` here would re-enter a
    /// half-initialized AppConfig. Callers log failures at their own site.
    @discardableResult
    static func createCacheDir(_ dir: URL) -> Bool {
        let fm = FileManager.default
        var isDir: ObjCBool = false
        if fm.fileExists(atPath: dir.path, isDirectory: &isDir), isDir.boolValue { return true }
        return (try? fm.createDirectory(at: dir, withIntermediateDirectories: true)) != nil
    }

    /// Switch the agent to the built-in royalty-free review library (15 tracks),
    /// used when signed in as the demo/review account. The bundled XML carries
    /// build-time absolute audio paths, so we rewrite each Location to the audio
    /// files inside THIS app bundle before pointing the parser at it. The real
    /// Rekordbox DB is never touched (DB source disabled).
    func activateReviewMode() {
        guard let resURL = Bundle.main.resourceURL else { return }
        let srcXML   = resURL.appendingPathComponent("review_library/rimeo_review.xml")
        let audioDir = resURL.appendingPathComponent("review_library/audio")
        guard let raw = try? String(contentsOf: srcXML, encoding: .utf8) else { return }

        // Rewrite Location="file://localhost/.../audio/<file>.mp3" → this bundle.
        var rewritten = raw
        if let re = try? NSRegularExpression(pattern: #"Location="file://localhost[^"]*?/([^"/]+\.mp3)""#) {
            let ns = NSMutableString(string: raw)
            for m in re.matches(in: raw, range: NSRange(location: 0, length: (raw as NSString).length)).reversed() {
                let fn  = (raw as NSString).substring(with: m.range(at: 1))
                let enc = audioDir.appendingPathComponent(fn).path
                    .addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? ""
                ns.replaceCharacters(in: m.range, with: "Location=\"file://localhost\(enc)\"")
            }
            rewritten = ns as String
        }

        let activeXML = baseDir.appendingPathComponent("review_active.xml")
        try? rewritten.write(to: activeXML, atomically: true, encoding: .utf8)

        queue.sync {
            xmlPath = activeXML.path
            reviewMode = true
        }
        // Persist so a restart stays in review mode; XML-only, never the real DB.
        updateEnvVar(key: "RIMEO_XML_PATH", value: activeXML.path)
        setXMLSourceEnabled(true)
        setDBSourceEnabled(false)
    }

    func setXMLPath(_ path: String) {
        queue.sync { xmlPath = path }
        updateEnvVar(key: "RIMEO_XML_PATH", value: path)
    }

    func setDBPath(_ path: String) {
        queue.sync { dbPath = path }
        updateEnvVar(key: "RIMEO_DB_PATH", value: path)
    }

    func setDBSourceEnabled(_ on: Bool) {
        queue.sync { dbSourceEnabled = on }
        updateEnvVar(key: "RIMEO_DB_ENABLED", value: on ? "1" : "0")
    }

    func setXMLSourceEnabled(_ on: Bool) {
        queue.sync { xmlSourceEnabled = on }
        updateEnvVar(key: "RIMEO_XML_ENABLED", value: on ? "1" : "0")
    }

    func localAgentURL() -> String {
        "http://\(getLocalIP()):\(port)"
    }

    var xmlExists: Bool {
        !xmlPath.isEmpty && FileManager.default.fileExists(atPath: xmlPath)
    }

    var dbExists: Bool {
        !dbPath.isEmpty && FileManager.default.fileExists(atPath: dbPath)
    }

    var hasAnyLibrarySource: Bool {
        xmlExists || dbExists
    }

    func applyCloudHeaders(to request: inout URLRequest, contentType: String? = nil) {
        if let contentType {
            request.setValue(contentType, forHTTPHeaderField: "Content-Type")
        }
        request.setValue(
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
            forHTTPHeaderField: "User-Agent"
        )
        request.setValue("application/json", forHTTPHeaderField: "Accept")
    }

    func getLocalIP() -> String {
        let sock = socket(AF_INET, SOCK_DGRAM, 0)
        guard sock >= 0 else { return "127.0.0.1" }
        defer { Darwin.close(sock) }

        var dst = sockaddr_in()
        dst.sin_family  = sa_family_t(AF_INET)
        dst.sin_port    = 80
        dst.sin_addr.s_addr = 0x08080808   // 8.8.8.8

        let ok = withUnsafePointer(to: &dst) {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                Darwin.connect(sock, $0, socklen_t(MemoryLayout<sockaddr_in>.size)) == 0
            }
        }
        guard ok else { return "127.0.0.1" }

        var local = sockaddr_in()
        var len   = socklen_t(MemoryLayout<sockaddr_in>.size)
        withUnsafeMutablePointer(to: &local) {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                getsockname(sock, $0, &len)
            }
        }
        var buf = [CChar](repeating: 0, count: Int(INET_ADDRSTRLEN))
        var inAddr = local.sin_addr
        inet_ntop(AF_INET, &inAddr, &buf, socklen_t(INET_ADDRSTRLEN))
        let ip = String(cString: buf)
        return ip.isEmpty ? "127.0.0.1" : ip
    }

    private static func makeDisplayVersion(version: String, buildNumber: String) -> String {
        let trimmed = buildNumber.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, trimmed.lowercased() != "dev" else { return version }
        return "\(version) (build \(trimmed))"
    }

    private static func detectDBPath() -> String {
        let home = FileManager.default.homeDirectoryForCurrentUser.path
        return "\(home)/Library/Pioneer/rekordbox/master.db"
    }

    private static func loadBuildInfo() -> (version: String, buildNumber: String, releaseTag: String) {
        let fallback = ("1.0", "dev", "v1.0-dev")
        let fm = FileManager.default
        let cwd = URL(fileURLWithPath: fm.currentDirectoryPath)
        let execDir = URL(fileURLWithPath: CommandLine.arguments[0]).deletingLastPathComponent()

        let searchRoots = [
            cwd,
            cwd.deletingLastPathComponent(),
            execDir,
            execDir.deletingLastPathComponent(),
            execDir.deletingLastPathComponent().deletingLastPathComponent(),
            Bundle.main.resourceURL,
        ]
        .compactMap { $0 }

        for root in searchRoots {
            let candidate = root.appendingPathComponent("build_info.py")
            guard let text = try? String(contentsOf: candidate, encoding: .utf8) else { continue }
            let version = match(in: text, pattern: #"VERSION\s*=\s*"([^"]+)""#) ?? fallback.0
            let build = match(in: text, pattern: #"BUILD_NUMBER\s*=\s*"([^"]+)""#) ?? fallback.1
            let tag = match(in: text, pattern: #"RELEASE_TAG\s*=\s*"([^"]+)""#) ?? fallback.2
            return (version, build, tag)
        }

        return fallback
    }

    private static func match(in text: String, pattern: String) -> String? {
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return nil }
        let range = NSRange(text.startIndex..<text.endIndex, in: text)
        guard let match = regex.firstMatch(in: text, range: range),
              match.numberOfRanges > 1,
              let valueRange = Range(match.range(at: 1), in: text) else { return nil }
        return String(text[valueRange])
    }

    private func updateEnvVar(key: String, value: String) {
        let envFile = baseDir.appendingPathComponent(".env")
        let existing = (try? String(contentsOf: envFile, encoding: .utf8)
            .components(separatedBy: .newlines)) ?? []
        var lines = existing.filter { !$0.trimmingCharacters(in: .whitespaces).hasPrefix("\(key)=") && !$0.isEmpty }
        lines.append("\(key)=\(value)")
        try? (lines.joined(separator: "\n") + "\n")
            .write(to: envFile, atomically: true, encoding: .utf8)
    }
}

File: windows_csharp/RimeoAgent/Config/AppConfig.cs
using System.Net.Sockets;
using System.Text.RegularExpressions;

namespace RimeoAgent.Config;

public sealed class AppConfig
{
    public static readonly AppConfig Shared = new();

    public const string AppName = "Rimeo Desktop Agent";
    public const string RimeoAppUrl = "https://rimeo.app";
    public const string GithubRepo  = "ilokhrimenko-lab/rimeo-agent";
    public const int    Port        = 8000;

    public string Version       { get; private set; } = "1.0";
    public string BuildNumber   { get; private set; } = "dev";
    public string ReleaseTag    { get; private set; } = "v1.0-dev";
    public string DisplayVersion { get; private set; } = "1.0";

    public string BaseDir      { get; }
    public string CacheDir     { get; }
    public string DataFile     { get; }
    public string LogFile      { get; }
    public string AnalysisFile { get; }
    public string AgentId      { get; private set; }

    private string _xmlPath = "";
    private string _dbPath  = "";
    private readonly object _lock = new();

    public string XmlPath { get { lock (_lock) return _xmlPath; } }
    public string DbPath  { get { lock (_lock) return _dbPath;  } }

    public bool XmlExists => !string.IsNullOrEmpty(XmlPath) && File.Exists(XmlPath);
    public bool DbExists  => !string.IsNullOrEmpty(DbPath)  && File.Exists(DbPath);
    public bool HasAnyLibrarySource => XmlExists || DbExists;

    private AppConfig()
    {
        var appData = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
        BaseDir      = Path.Combine(appData, "RimeoAgent");
        CacheDir     = Path.Combine(BaseDir, "cache");
        DataFile     = Path.Combine(BaseDir, "rimo_data.json");
        LogFile      = Path.Combine(BaseDir, "agent.log");
        AnalysisFile = Path.Combine(BaseDir, "analysis_data.json");

        Directory.CreateDirectory(BaseDir);
        Directory.CreateDirectory(CacheDir);

        // Persistent agent ID
        var idFile = Path.Combine(BaseDir, "agent_id");
        if (File.Exists(idFile))
        {
            AgentId = File.ReadAllText(idFile).Trim();
            if (string.IsNullOrEmpty(AgentId)) AgentId = GenerateAndSaveId(idFile);
        }
        else
        {
            AgentId = GenerateAndSaveId(idFile);
        }

        LoadBuildInfo();

        // Pioneer rekordbox master.db default path
        _dbPath = DetectDbPath();

        // Load .env for overrides
        var envFile = Path.Combine(BaseDir, ".env");
        if (File.Exists(envFile))
        {
            foreach (var line in File.ReadAllLines(envFile))
            {
                var parts = line.Split('=', 2);
                if (parts.Length < 2) continue;
                var key = parts[0].Trim();
                var val = parts[1].Trim();
                if (key == "RIMEO_XML_PATH") _xmlPath = val;
                else if (key == "RIMEO_DB_PATH" && !string.IsNullOrEmpty(val)) _dbPath = val;
            }
        }
    }

    public void SetXmlPath(string path)
    {
        lock (_lock) { _xmlPath = path; }
        UpdateEnvVar("RIMEO_XML_PATH", path);
    }

    public void SetDbPath(string path)
    {
        lock (_lock) { _dbPath = path; }
        UpdateEnvVar("RIMEO_DB_PATH", path);
    }

    public string GetLocalIp()
    {
        try
        {
            using var socket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
            socket.Connect("8.8.8.8", 80);
            return (socket.LocalEndPoint as System.Net.IPEndPoint)?.Address.ToString() ?? "127.0.0.1";
        }
        catch
        {
            return "127.0.0.1";
        }
    }

    public string LocalAgentUrl() => $"http://{GetLocalIp()}:{Port}";

    public void ApplyCloudHeaders(HttpRequestMessage req, string? contentType = null)
    {
        if (contentType != null)
            req.Content!.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue(contentType);
        req.Headers.TryAddWithoutValidation("User-Agent",
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36");
        req.Headers.Accept.ParseAdd("application/json");
    }

    private void LoadBuildInfo()
    {
        var searchDirs = new[]
        {
            AppContext.BaseDirectory,
            Path.GetDirectoryName(AppContext.BaseDirectory) ?? "",
            Path.GetDirectoryName(Environment.ProcessPath ?? "") ?? "",
        };

        foreach (var dir in searchDirs.Where(d => !string.IsNullOrEmpty(d)))
        {
            var candidate = Path.Combine(dir, "build_info.py");
            if (!File.Exists(candidate)) continue;
            var text = File.ReadAllText(candidate);
            Version     = ExtractPyStr(text, "VERSION")     ?? "1.0";
            BuildNumber = ExtractPyStr(text, "BUILD_NUMBER") ?? "dev";
            ReleaseTag  = ExtractPyStr(text, "RELEASE_TAG") ?? "v1.0-dev";
            var b = BuildNumber.Trim();
            DisplayVersion = (string.IsNullOrEmpty(b) || b.ToLower() == "dev")
                ? Version
                : $"{Version} (build {b})";
            return;
        }
    }

    private static string? ExtractPyStr(string text, string key)
    {
        var m = Regex.Match(text, $@"{key}\s*=\s*""([^""]+)""");
        return m.Success ? m.Groups[1].Value : null;
    }

    private static string DetectDbPath()
    {
        var appData = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
        var path = Path.Combine(appData, "Pioneer", "rekordbox", "master.db");
        return File.Exists(path) ? path : "";
    }

    private static string GenerateAndSaveId(string idFile)
    {
        var id = Guid.NewGuid().ToString().ToUpper();
        File.WriteAllText(idFile, id);
        return id;
    }

    private void UpdateEnvVar(string key, string value)
    {
        var envFile = Path.Combine(BaseDir, ".env");
        var lines = File.Exists(envFile)
            ? File.ReadAllLines(envFile).Where(l => !l.TrimStart().StartsWith($"{key}=")).ToList()
            : new List<string>();
        lines.Add($"{key}={value}");
        File.WriteAllLines(envFile, lines);
    }
}
